### ISSUE TITLE
Don't run composer as root

### ISSUE THREAD

**Issue by oxc** [tohuwabohu/puppet-roundcube#38]
I've recently started getting errors on puppet run:

```
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns: Do not run Composer as root/super user! See https://getcomposer.org/root for details
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:                                                                                   
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns: In PluginManager.php line 769:
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:                                                                                   
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:   roundcube/plugin-installer contains a Composer plugin which is blocked by y
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:   our allow-plugins config. You may add it to the list if you consider it saf
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:   e.
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:   You can run "composer config --no-plugins allow-plugins.roundcube/plugin-in
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:   staller [true|false]" to enable it (true) or disable it explicitly and supp
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:   ress this exception (false)                                                     
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:   See https://getcomposer.org/allow-plugins
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns:
Notice: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns: install [--prefer-source] [--prefer-dist] [--prefer-install PREFER-INSTALL] [--dry-run] [--dev] [--no-suggest] [--no-dev] [--no-autoloader] [--no-progress] [--no-install] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--apcu-autoloader-prefix APCU-AUTOLOADER-PREFIX] [--ignore-platform-req IGNORE-PLATFORM-REQ] [--ignore-platform-reqs] [--] [<packages>...]
Error: 'composer install --no-dev' returned 1 instead of one of [0]
Error: /Stage[main]/Roundcube::Install/Exec[composer install --no-dev]/returns: change from 'notrun' to ['0'] failed: 'composer install --no-dev' returned 1 instead of one of [0] (corrective)
```

Would it be possible to run the composer commands with a different user. Probably something like www-data or nobody, and preferably configurable.

**Comment by oxc** [comment]
The issue seems in fact to be caused by the unallowed plugin. However, for reasons mentioned in https://getcomposer.org/root, it is desirable to still have a way to not run composer as root, so I'll keep this issue as it is.

**Comment by tohuwabohu** [comment]
> Would it be possible to run the composer commands with a different user. Probably something like www-data or nobody, and preferably configurable.

Yes, I agree it would definitely help to not run the install process as root. Too bad that the linked documentation doesn't provide a clear path forward how to fix this, besides disabling plugins and scripts. I don't know if roundcube is currently using this or if that would be a suitable option.

I think running as www-data is not desirable either as it would allow attackers to persist malicious code by changing the underlying php files. Ideally the roundcube installation remains read-only for the www process.

Using nobody could be a good way out, unfortunately I haven't found anything on the Internet which indicates this will work successfully.

**Comment by oxc** [comment]
> I think running as www-data is not desirable either as it would allow attackers to persist malicious code by changing the underlying php files. Ideally the roundcube installation remains read-only for the www process.

I don't think this is something to worry about, because the attacker in question here is the one providing (some of) the php files anyway.
If the plugins installed by composer are compromised, you have a problem anyway, but at least it only compromises your roundcube install, and not immediately the whole system.

It think it should be sufficient to let the module user decide which user to use. For many people using www-data is sufficiently secure (think about self-updating wordpress etc.), if the admin has higher security requirements they can specify their own user whose umask is readable by the www user.

**Comment by tohuwabohu** [comment]
True!

My worry about using www-data was not regarding the install process but more when running the applicaiton: let's say roundcube ships a vulnerability that allows someone to upload attachements and the attacker could control the path where the file is stored on server side, then this could potentially be used to replace one of the php files in the installation folder.

I agree that even just exposing the user used to run the scripts would offer enough flexibility for most users. I'll see if I get to it in the next couple of days. If you need this sooner I'm more than happy to review a pull request :smile:

**Comment by tohuwabohu** [comment]
I've released version 3.6.0 to the forge containing a `composer_user` parameter to control the user.
